docs/addmember.1
.TH ADDMEMBER 1 "December 16, 2007"
.SH NAME
addmember \- add club members to the directory
.SH SYNOPSIS
.B addmember
userid name [ program ]
.SH DESCRIPTION
.B Addmember
performs all tasks necessary for creation of a new CSC member. It creates
an LDAP entry, Kerberos principal, and home directory for the new member.
It does NOT register the new member for any terms. This must be done after
the member is created.
.SH SEE ALSO
.BR ceo (1).
.SH AUTHOR
Michael Spang <mspang@csclub.uwaterloo.ca>


debian/changelog
ceo (0.4.1) stable testing; urgency=low

  * Minor fixes

 -- Michael Spang <mspang@uwaterloo.ca>  Wed, 12 Dec 2007 03:40:17 -0500

ceo (0.4.0) stable testing; urgency=low

  * New release

 -- Michael Spang <mspang@uwaterloo.ca>  Wed, 12 Dec 2007 03:07:05 -0500

ceo (0.3.9) stable testing; urgency=low

  * New release

 -- Michael Spang <mspang@uwaterloo.ca>  Mon, 10 Dec 2007 03:56:06 -0500

ceo (0.3.3) stable testing; urgency=low

  * Add club and group modify page
  * Add sasl support
  * Complete group and position management
  * Remove ceo-old
  * Fix bugs

 -- David Bartley <dtbartle@csclub.uwaterloo.ca>  Wed, 21 Nov 2007 20:56:14 -0500

ceo (0.3.2) unstable; urgency=low

  [ Michael Spang ]
  * Fix CEO group add for rfc2307bis

  [ David Bartley ]
  * Add 'search by group'
  * Lookup name and program based on uwdir id
  * Add group and position management

 -- Michael Spang <mspang@uwaterloo.ca>  Wed, 21 Nov 2007 17:21:40 -0500

ceo (0.3.1) unstable; urgency=low

  * addhomedir: invalidate nscd tables
  * ceo-urwid: add create club account menuitem
  * Add urwid to dependencies

 -- Michael Spang <mspang@uwaterloo.ca>  Fri,  5 Oct 2007 10:16:41 -0400

ceo (0.3.0) unstable; urgency=low

  * Add experimental urwid-based GUI
  * Rip out studentid support
  * Unbreak termusers in ceoquery
  * Increase widths of UI windows
  * PgSQL to LDAP transition

 -- Michael Spang <mspang@uwaterloo.ca>  Tue, 25 Sep 2007 04:00:10 -0400

ceo (0.2.4) unstable; urgency=low

  * Added csc.schema.
  * Vim-style keybindings for CEO menus.
  * Bug fix: call setreuid(euid, euid) in csc-chfn and csc-chsh.
  * Bug fix: run less in "secure" mode.
  * Renamed package to ceo.

 -- Michael Spang <mspang@uwaterloo.ca>  Mon, 28 May 2007 02:05:28 -0400

csc (0.2.3) unstable; urgency=low

  * Added "ceoquery", a utility to retrieve lists of members and users.
  * Added "csc-chsh" and "csc-chfn" utilities.
  * Bug fix: build_gecos() did not include enough commas between fields.
  * Member attributes are now added to LDAP as well as the PgSQL database.

 -- Michael Spang <mspang@uwaterloo.ca>  Sun, 18 Feb 2007 21:35:28 -0500

csc (0.2.2) unstable; urgency=low

  * Added "addhomedir", a utility to create home directories for new users.
  * Bug fix: CEO still referenced an exception that changed name in 0.2.
  * Documentation updates.

 -- Michael Spang <mspang@uwaterloo.ca>  Mon, 29 Jan 2007 01:47:31 -0500

csc (0.2.1) unstable; urgency=low

  * Documentation updates only
  * Added docs/GIT-HOWTO and docs/INSTALLING

 -- Michael Spang <mspang@uwaterloo.ca>  Sun, 28 Jan 2007 01:24:37 -0500

csc (0.2) unstable; urgency=low

  * Tests added to most Python modules.
  * Split configuration files.
  * Added maintainer scripts to manage permissions during install and purge.
  * Added functions for use by tools planned for next release (chfn, etc).
  * Added support for account "repair", which will recreate LDAP entries
    and principals if necessary.
  * The recreate account menu option in ceo is now active.
  * Replaced instances of "== None" and "!= None" with "is None" and
    "is not None", respectively (thanks to: Nick Guenther).
  * Renamed terms.valid() to terms.validate() (thanks to: Nick Guenther).

 -- Michael Spang <mspang@uwaterloo.ca>  Fri, 26 Jan 2007 20:10:14 -0500

csc (0.1) unstable; urgency=low

  * Initial Release.

 -- Michael Spang <mspang@uwaterloo.ca>  Thu, 28 Dec 2006 04:07:03 -0500



src/addmember.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/acl.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <ctype.h>
#include <pwd.h>
#include <grp.h>
#include <errno.h>
#include <libgen.h>
#include <syslog.h>

#include "util.h"
#include "common.h"
#include "config.h"
#include "ldap.h"
#include "krb5.h"
#include "kadm.h"
#include "addhomedir.h"

char *prog = NULL;
char *user = NULL;
int privileged = 0;

static int force = 0;
static int no_notify = 0;

static int use_stdin = 0;

static char *name = NULL;
static char *userid = NULL;
static char *program = NULL;
static char password[1024];

static struct option opts[] = {
    { "force", 0, NULL, 'f' },
    { "no-notify", 0, NULL, 'q' },
    { "stdin", 0, NULL, 's' },
    { NULL, 0, NULL, '\0' },
};

static void usage() {
    fprintf(stderr, "Usage: %s userid realname [program]\n", prog);
    exit(2);
}

int addmember() {
    int krb_ok, user_ok, group_ok, home_ok, quota_ok;
    int id;
    char homedir[1024];
    acl_t acl = NULL, dacl = NULL;

    logmsg("adding uid=%s cn=%s program=%s by %s", userid, name, program, user);

    if (setreuid(0, 0))
        fatalpe("setreuid");

    if (!force && getpwnam(userid) != NULL)
        deny("user %s already exists", userid);

    snprintf(homedir, sizeof(homedir), "%s/%s", member_home, userid);

    acl = acl_from_text(member_home_acl);
    if (acl == NULL)
        fatalpe("Unable to parse member_home_acl");

    if (*member_home_acl) {
        dacl = acl_from_text(member_home_dacl);
        if (dacl == NULL)
            fatalpe("Unable to parse member_home_dacl");
    }

    if (ceo_read_password(password, sizeof(password), use_stdin))
        return 1;

    ceo_krb5_init();
    ceo_ldap_init();
    ceo_kadm_init();

    if (ceo_user_exists(userid))
        deny("user %s already exists in LDAP", userid);

    if ((id = ceo_new_uid(member_min_id, member_max_id)) <= 0)
        fatal("no available uids in range [%d, %d]", member_min_id, member_max_id);

    krb_ok = ceo_del_princ(userid);
    krb_ok = krb_ok || ceo_add_princ(userid, password);
    if (!krb_ok)
        logmsg("successfully created principal for %s", userid);

    user_ok = krb_ok || ceo_add_user(userid, users_base, "member", name, homedir,
            member_shell, id, "program", program, NULL);
    if (!user_ok)
        logmsg("successfully created account for %s", userid);

    group_ok = user_ok || ceo_add_group(userid, groups_base, id);
    if (!group_ok)
        logmsg("successfully created group for %s", userid);

    home_ok = user_ok || ceo_create_home(homedir, id, id, acl, dacl);
    if (!home_ok)
        logmsg("successfully created home directory for %s", userid);

    quota_ok = user_ok || ceo_set_quota(quota_prototype, id);
    if (!quota_ok)
        logmsg("successfully set quota for %s", userid);

    logmsg("done uid=%s", userid);

    if (!no_notify && !user_ok) {
        int pid;
        int hkp[2];
        FILE *hkf;
        int status;

        if (pipe(hkp))
            errorpe("pipe");

        fflush(stdout);
        fflush(stderr);

        pid = fork();

        if (!pid) {
            fclose(stdout);
            fclose(stderr);
            close(hkp[1]);
            dup2(hkp[0], 0);
            exit(execl(notify_hook, notify_hook, prog, user, userid, name, program, NULL));
        }

        hkf = fdopen(hkp[1], "w");

        if (group_ok)
            fprintf(hkf, "failed to create group\n");
        if (home_ok)
            fprintf(hkf, "failed to create home directory\n");
        if (quota_ok)
            fprintf(hkf, "failed to set quota\n");
        if (!group_ok && !home_ok && !quota_ok)
            fprintf(hkf, "all failures went undetected\n");

        fclose(hkf);

        waitpid(pid, &status, 0);

        if (WIFEXITED(status) && WEXITSTATUS(status))
            logmsg("hook %s exited with status %d", notify_hook, WEXITSTATUS(status));
        else if (WIFSIGNALED(status))
            logmsg("hook %s killed by signal %d", notify_hook, WTERMSIG(status));
    }

    ceo_kadm_cleanup();
    ceo_ldap_cleanup();
    ceo_krb5_cleanup();

    return krb_ok || user_ok || group_ok || home_ok || quota_ok;
}

int main(int argc, char *argv[]) {
    int opt;

    openlog(prog, 0, LOG_AUTHPRIV);

    configure();

    prog = basename(argv[0]);
    user = ceo_get_user();
    privileged = ceo_get_privileged();

    while ((opt = getopt_long(argc, argv, "", opts, NULL)) != -1) {
        switch (opt) {
            case 's':
                use_stdin = 1;
                break;
            case 'f':
                if (!privileged)
                    deny("not privileged enough to force");
                force = 1;
                break;
            case 'q':
                if (!privileged)
                    deny("not privileged enough to suppress notifications");
                no_notify = 1;
                break;
            case '?':
                usage();
                break;
            default:
                fatal("error parsing arguments");
        }
    }

    if (argc - optind != 2 && argc - optind != 3)
        usage();

    userid = argv[optind++];
    name = argv[optind++];

    if (argc - optind)
        program = argv[optind++];

    return addmember();
}
